File: Leetcode/637_Average_of_Levels_in_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>

class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        using namespace std;
        
        queue<pair<TreeNode*, int>> q;
        vector<double> averages;
        vector<double> nums;
        q.push(make_pair(root, 1));
        int prev = 1;
        
        while(!q.empty())
        {
            TreeNode* current = q.front().first;
            int depth = q.front().second;
            q.pop();
            if(prev != depth)
            {
                double sum = 0;
                for(auto num: nums)
                {
                    sum += num;
                }
                averages.push_back(sum / nums.size());
                nums.clear();
                prev = depth;
            }
            nums.push_back(current->val);
            
            if(current->left != nullptr)
            {
                q.push(make_pair(current->left, depth + 1));
            }
            if(current->right != nullptr)
            {
                q.push(make_pair(current->right, depth + 1));
            }
        }
        
        double sum = 0;
        for(auto num: nums)
        {
            sum += num;
        }
        averages.push_back(sum / nums.size());
        
        return averages;
    }
};

File: Leetcode/0637/637_Average_of_Levels_in_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function (root) {
    let result = [];
    traversal(root, 0, result);
    let answer = [];
    return calculateAverage(result, answer);

};

var traversal = function (root, depth, result) {
    if (root) {
        traversal(root.left, depth + 1, result);
        result.push({ val: root.val, depth });
        traversal(root.right, depth + 1, result);
    }
}

var calculateAverage = function (result, answer) {
    for (let i = 0; i < result.length; i++) {
        if (!answer[result[i].depth]) {
            answer[result[i].depth] = { sum: result[i].val, counts: 1 };
            continue;
        }
        let target = answer[result[i].depth];
        answer[result[i].depth] = { sum: parseInt(target.sum) + result[i].val, counts: parseInt(target.counts) + 1 }
    }

    return answer.reduce((acc, cur) => {
        acc.push(cur.sum / cur.counts)
        return acc;
    }, [])
}